src/main/scala/beam/agentsim/agents/vehicles/BeamVehicle.scala

```diff
@@ -1,46 +1,122 @@
 package beam.agentsim.agents.vehicles
 
-import beam.agentsim.agents.BeamAgent
+import akka.actor.{Actor, ActorLogging, ActorRef}
+import akka.pattern._
+import beam.agentsim.Resource
 import beam.agentsim.agents.BeamAgent.BeamAgentData
 import beam.agentsim.events.SpaceTime
-import org.matsim.api.core.v01.{Id, Identifiable}
+import beam.agentsim.events.resources.vehicle.GetVehicleLocationEvent
+import beam.sim.config.BeamConfig
+import org.matsim.api.core.v01.Id
+import org.matsim.utils.objectattributes.attributable.Attributes
+import org.matsim.vehicles.{Vehicle, VehicleType}
+import akka.pattern.pipe
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
 
 /**
   * @author dserdiuk
   */
 
 abstract class Dimension
 
-case class VehicleData(dim: Dimension, trajectory: Trajectory) extends BeamAgentData
+object VehicleData {
+  implicit def vehicle2vehicleData(vehicle: Vehicle): VehicleData = {
+    val vdata = VehicleDataImpl(vehicle.getType.getDescription,
+      vehicle.getClass.getName, vehicle, new Attributes())
+    vdata
+  }
+}
+trait VehicleData extends BeamAgentData with Vehicle {
+  /**
+    * It's pretty general name of type of vehicle.
+    * It could be a model name of particular brand as well as vehicle class: sedan, truck, bus etc.
+    * The key point this type need to be unique
+    * @return
+    */
+  def vehicleTypeName: String
 
+  /**
+    * MATSim vehicle vehicle implementation class
+    * @return
+    */
+  def vehicleClassName: String
+}
+
+object BeamVehicle {
 
-trait BeamVehicle[T] extends Identifiable[T]{
+  def energyByType(vehicleTypeId: Id[VehicleType]): Double = {
```

Comment:
Ok, I see now, it's energy per unit length. Can you rename to "energyPerUnitByType"?